src/librustc_parse/parser/expr.rs

```diff
@@ -621,7 +621,44 @@ impl<'a> Parser<'a> {
                     }
                 }
             }
-        }
+        };
+
+        // Disallow postfix operators such as `.`, `?` or index (`[]`) after casts.
+        // Parses the postfix operator and emits an error.
+        let expr = type_result?;
+        let span = expr.span;
+
+        // The resulting parse tree for `&x as T[0]` has a precedence of `((&x) as T)[0]`.
+        let with_postfix = self.parse_dot_or_call_expr_with_(expr, span)?;
+        if !matches!(with_postfix.kind, ExprKind::Cast(_, _)) {
```

Comment:
What is this condition for? It's either unnecessary (always true) or incorrect.
We must always report an error if we parsed something postfix after the cast/ascription expression, but here it looks like we may accept things like `expr as Type1.field as Type2` without an error.

(Also `ExprCast` and `ExprType` should always do the same thing synctactically, e.g. if `expr as Type1 as Type2` work, then `expr as Type1: Type2` should also work.)

Reply:
I believe we call too far down into the parse chain to ever parse another `as` expression here. If we've read the code correctly, `parse_dot_or_call_expr_with_` will only ever parse things starting with `.`, `[` or `(` after an existing expression. We would have had to call into `parse_assoc_expr` in order to parse another `as` expression.

On parsing any expression, I think it goes something like - leaving out intermediate methods like all the ones which are `_with ` and `_`:

- `parse_expr` -> parses outer attributes, forwards to `parse_assoc_expr`
- `parse_assoc_expr` -> parses `*,+,||,&&` as well as any other binary operators and `as`. Calls helpers like `parse_assoc_expr_cast` when it encounters these. Calls into `parse_prefix_expr` to parse the left hand side of the operator.
- `parse_assoc_expr_*` (`assoc`, `ascribe`, `parse_range_expr`): these parse the thing they're named for, using the lhs parse from `parse_assoc_expr`
- `parse_prefix_expr`: this parses prefix expressions (`!`,`-`,`*`,`&`) then calls: `parse_bottom_expr` and  `parse_dot_or_call_expr_with` (passing in argument from `parse_bottom_expr`)
- `parse_bottom_expr`: parses single things like return expressions, parenthesized expressions, etc.
- `parse_dot_or_call_expr_with`: parses function calls, index operators, `?` operators and field operators, and applies those to the result passed in (usually from `parse_bottom_expr`)

By jumping directly to `parse_dot_or_call_expr_with`, we shouldn't be able to parse any associative operators, including `as X`, and only parse postfix things like indexing, function calls, and and field access/method calls/.await.

Do you think it's worth adding a comment along these lines explaining why we're doing this?

Reply:
It's correct that `parse_dot_or_call_expr_with_` cannot possibly result in `::Cast` or `::Type` again and will give back the expression passed to it in those cases. However, this is also relying on facts on the ground that may change with time. To prevent such accidents in the future, I think we should make some enhancements:

- Let's make sure there's a test for `expr :/as Type1 :/as Type2` (4 cases, test all of them) and that these are parse failures.
- When `parse_dot_or_call_expr_with_` passes back the expression, it does not reallocate. This means that you should be able to hash the address before the call and after and they should be the same. So this can be added as an `||`-ed condition after the `!matches(...)`. This shouldn't change the semantics, but it does add a measure of robustness.

> But this introduces new test failures, so we've clearly done something wrong. Maybe this is it?

The failures come from not anticipating that this function is also used for type ascriptions (as noted by @petrochenkov), e.g. `foo: Vec<u8>`. (See the function right below this one for how that happens.)

Reply:
Makes sense that we want to make it more reliable- we've added in address hashing, or at least what I think is good address hashing.

As for `expr :/as Type1 :/as Type2`, I've added tests, but they don't fail. Looking at how this behaves prior to these changes, I don't think these are expected to fail? `x as T1 as T2` is valid today. This doesn't change that, it'll still just parse `x as T1`, then fail to see any postfix, then go back up to `parse_assoc_expr` and parse the second `as T2` (or `: T2`).

Given that we test the expr pointer hash now, if `parse_dot_or_call_expr_with_` ever changes to start parsing type ascriptions or casts, then these tests _should_ start to fail because we'll trigger our error message. I think this is OK?

---

> The failures come from not anticipating that this function is also used for type ascriptions (as noted by @petrochenkov), e.g. foo: Vec<u8>. (See the function right below this one for how that happens.)

We've gone through and fixed this, I think.